std/arith.asm

```diff
@@ -17,18 +23,174 @@ machine Arith(CLK32_31, operation_id){
     
     // Performs elliptic curve addition of points (x1, y2) and (x2, y2).
     // Operation ID is 10 = 0b1010, i.e., we activate equations 1, 3, and 4.
-    // TODO: Witgen doesn't work for that yet.
     operation ec_add<10> x1c[0], x1c[1], x1c[2], x1c[3], x1c[4], x1c[5], x1c[6], x1c[7], y1c[0], y1c[1], y1c[2], y1c[3], y1c[4], y1c[5], y1c[6], y1c[7], x2c[0], x2c[1], x2c[2], x2c[3], x2c[4], x2c[5], x2c[6], x2c[7], y2c[0], y2c[1], y2c[2], y2c[3], y2c[4], y2c[5], y2c[6], y2c[7] -> x3c[0], x3c[1], x3c[2], x3c[3], x3c[4], x3c[5], x3c[6], x3c[7], y3c[0], y3c[1], y3c[2], y3c[3], y3c[4], y3c[5], y3c[6], y3c[7];
     
     // Performs elliptic curve doubling of point (x1, y2).
     // Operation ID is 12 = 0b1100, i.e., we activate equations 2, 3, and 4.
-    // TODO: Witgen doesn't work for that yet.
     operation ec_double<12> x1c[0], x1c[1], x1c[2], x1c[3], x1c[4], x1c[5], x1c[6], x1c[7], y1c[0], y1c[1], y1c[2], y1c[3], y1c[4], y1c[5], y1c[6], y1c[7] -> x3c[0], x3c[1], x3c[2], x3c[3], x3c[4], x3c[5], x3c[6], x3c[7], y3c[0], y3c[1], y3c[2], y3c[3], y3c[4], y3c[5], y3c[6], y3c[7];
     
     let BYTE: col = |i| i & 0xff;
     let BYTE2: col = |i| i & 0xffff;
 
-    pol commit x1[16], y1[16], x2[16], y2[16], x3[16], y3[16], s[16], q0[16], q1[16], q2[16];
+    let secp_modulus = 0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffefffffc2f;
+
+    let inverse: int -> int = |x| ff::inverse(x, secp_modulus);
+    let add = |x, y| ff::add(x, y, secp_modulus);
+    let sub = |x, y| ff::sub(x, y, secp_modulus);
+    let mul = |x, y| ff::mul(x, y, secp_modulus);
+    let div = |x, y| ff::div(x, y, secp_modulus);
+
+    pol commit x1[16], y1[16], x2[16], y2[16], x3[16], y3[16];
+
+    // Selects the ith limb of x (little endian)
+    // Note that the most significant limb can be up to 32 bits; all others are 16 bits.
+    let select_limb = |x, i| fe((x >> (i * 16)) & if i < 15 { 0xffff } else { 0xffffffff });
+
+    let s_for_eq1 = |x1, y1, x2, y2| div(sub(y2, y1), sub(x2, x1));
```

Comment:
Should we not make this
```suggestion
    let s_for_eq1 = || div(sub(y2_int(), y1_int()), sub(x2_int(), x1_int()));
```

Reply:
Hmm, sure, but for other functions like `s_for_eq2`, this would mean computing the same number twice. And doing it only for `s_for_eq1` is inconsistent. So I'd leave it like this for now and revisit once we have `let` statements.